jetty-server/src/main/java/org/eclipse/jetty/server/OptionalConnectionFactory.java

```diff
@@ -0,0 +1,225 @@
+//  Copyright (c) 1995-2020 Mort Bay Consulting Pty Ltd and others.
+//  ------------------------------------------------------------------------
+//  All rights reserved. This program and the accompanying materials
+//  are made available under the terms of the Eclipse Public License v1.0
+//  and Apache License v2.0 which accompanies this distribution.
+//
+//      The Eclipse Public License is available at
+//      http://www.eclipse.org/legal/epl-v10.html
+//
+//      The Apache License v2.0 is available at
+//      http://www.opensource.org/licenses/apache2.0.php
+//
+//  You may elect to redistribute this code under either of these licenses.
+//  ========================================================================
+//
+
+package org.eclipse.jetty.server;
+
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.List;
+
+import org.eclipse.jetty.io.AbstractConnection;
+import org.eclipse.jetty.io.Connection;
+import org.eclipse.jetty.io.EndPoint;
+import org.eclipse.jetty.server.SensingConnectionFactory.SensingOutcome;
+import org.eclipse.jetty.util.BufferUtil;
+import org.eclipse.jetty.util.component.LifeCycle;
+import org.eclipse.jetty.util.log.Log;
+import org.eclipse.jetty.util.log.Logger;
+
+/**
+ * A {@link ConnectionFactory} combining multiple {@link SensingConnectionFactory} instances that will upgrade to
+ * the first one recognizing the bytes in the buffer.
+ */
+public class OptionalConnectionFactory extends AbstractConnectionFactory
+{
+    private static final Logger LOG = Log.getLogger(OptionalConnectionFactory.class);
+
+    private final String _nextProtocol;
+    private final List<SensingConnectionFactory> _sensingConnectionFactories;
+
+    public OptionalConnectionFactory(SensingConnectionFactory... sensingConnectionFactories)
+    {
+        this(null, sensingConnectionFactories);
+    }
+
+    public OptionalConnectionFactory(String nextProtocol, SensingConnectionFactory... sensingConnectionFactories)
+    {
+        super(toProtocolString(sensingConnectionFactories));
+        _nextProtocol = nextProtocol;
+        _sensingConnectionFactories = Arrays.asList(sensingConnectionFactories);
+    }
+
+    static String toProtocolString(SensingConnectionFactory... sensingConnectionFactories)
+    {
+        StringBuilder sb = new StringBuilder();
+        for (SensingConnectionFactory ccf : sensingConnectionFactories)
+        {
+            String protocol = ccf.getProtocol();
+            sb.append(protocol).append('|');
+        }
+        sb.deleteCharAt(sb.length() - 1);
+        return sb.toString();
+    }
+
+    @Override
+    protected void doStart() throws Exception
+    {
+        super.doStart();
+
+        for (SensingConnectionFactory connectionFactory : _sensingConnectionFactories)
+        {
+            if (connectionFactory instanceof LifeCycle)
+                ((LifeCycle)connectionFactory).start();
+        }
```

Comment:
Since `AbstractConnectionFactory` _is-a_ `ContainerLifeCycle`, you don't want to manage the lifecycle directly in the way above.

You just add the children as beans in the constructor, and `ContainerLifeCycle` will take care of starting and stopping them.

Reply:
That makes sense, let me modify this, thanks!